### PR TITLE
fix setup call while testing

### LINKED ISSUE #2620: Only invoke setup() once, not in both trainer.fit() and trainer.test()
## 🚀 Feature

Only invoke `def setup(self, step: str)` when calling `trainer.test(net)` if this has not been called before (e.g. `trainer.fit(net)`).

### Motivation

The `setup` function is [described in the docs](https://pytorch-lightning.readthedocs.io/en/latest/introduction_guide.html#models-defined-by-data) as: "use setup to do splits, and build your model internals".
Therefore, I wrote code that does the train-val-test function and some DataFrame labels transformation (e.g. label to one_hot) in this function.
A pretty common pattern is the following:
```python
trainer.fit(net)
trainer.test(net)
```
Contrary to what I expected, I saw from my debug output that `setup` was invoked twice.
This is a **waste of computational resources**, and since I did the train-val split randomly, I do not have access to the indices that were used in either step (and possibly other issues such as the label transformation re-ordering the columns of which number represent which label).

Current situation, train and val step use same `setup`, test step uses another invoke of `setup`
I assume that it is more common that the train-val and test step of the trainer use the same setup code, than that setup does something special for only test (and not val).
As Lightning works by giving sensible defaults, and allowing you to hack at anything you want, the logic should be that `setup` should only be invoked once, and allow for a way to specify a special test `setup` function.

### Pitch

Have the trainer keep track of whether `setup()` has been invoked or not, so `setup()` can be skipped in `trainer.test(net)` if it was already invoked in `trainer.fit(net)`. This way the common use case will benefit from less computation power and is more in line what is expected of the magic of Lightning.

I'm not sure what would be the best approach for users to set a separate setup call for testing.
Maybe something like: `trainer.test(net, always_invoke_setup=True)`?

### Alternatives

Include some custom logic that checks if data has been initialized:
```python
def setup(self, step: str):
    if self.data is None:
        # setup code
    else:
        # do nothing
```

### Additional context

Ideas formed by discussing this issue on the pytorch-lightning SLACK in the questions channel. Thanks goed to the people who replied.

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #2620 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 1]
```diff
@@ -506,9 +506,10 @@ def ddp_train(self, process_idx, q, model, is_master=False, proc_offset=0):
         model.init_ddp_connection(self.global_rank, self.world_size, self.is_slurm_managing_tasks)
 
         # call setup after the ddp process has connected
-        self.setup('fit')
-        if self.is_function_implemented('setup', model):
-            model.setup('fit')
+        if not self.testing:
+            self.setup('fit')
+            if self.is_function_implemented('setup', model):
```
Comment (awaelchli) [inline pytorch_lightning/trainer/distrib_data_parallel.py:511]: why do we need this check for is_implemented? Can we not just call the the method unconditionally, and inheritance will take care of the rest?

Comment (rohitgr7) [reply pytorch_lightning/trainer/distrib_data_parallel.py:511]: You mean why `if self.is_function_implemented('setup', model):` is used or why `if self.is_function_implemented('setup', model):` is used within `if not self.testing:`?

Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_data_parallel.py:511]: The former. I suggest to drop the if statement and just call model.setup(). The if condition, as I see it, will always be True.

Comment (rohitgr7) [reply pytorch_lightning/trainer/distrib_data_parallel.py:511]: oh yeah! I got it confused with `is_overridden`. Yeah, it's not required.

**Comment by rohitgr7** [comment]
Quick question: As of now, it is calling `trainer.setup('test')` and `model.setup('test')` outside mp process only once for ddp or tpu (8 cores) training or horovod. Shouldn't the testing setup be called within mp process too to ensure ddp or any multi-device testing?

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/distrib_data_parallel.py]: There are more such places where this is done but not required [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/7b4db3045dcc9e6bb0b66e409b25bb2c7fa378f0/pytorch_lightning/trainer/trainer.py#L942), [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/7b4db3045dcc9e6bb0b66e409b25bb2c7fa378f0/pytorch_lightning/trainer/trainer.py#L1054), [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/7b4db3045dcc9e6bb0b66e409b25bb2c7fa378f0/pytorch_lightning/trainer/trainer.py#L1058) and many more. Should I fix them in this PR?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1306,8 +1306,7 @@ def test(
 
     def __test_using_best_weights(self, ckpt_path, test_dataloaders):
         model = self.get_model()
-        if self.is_function_implemented('setup', model):
-            model.setup('test')
+        model.setup('test')
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/trainer.py:1309]: Quick question: As of now, it is calling `trainer.setup('test')` and `model.setup('test')` outside mp process only once for ddp or tpu (8 cores) training or horovod. Shouldn't the testing setup be called within mp process too to ensure ddp or any multi-device testing?

Comment (williamFalcon) [reply pytorch_lightning/trainer/trainer.py:1309]: setup for anything DDP should happen inside spawn

Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1309]: should I move it there then for all?

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/distrib_data_parallel.py]: @awaelchli

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:1309]: @awaelchli Should I make the changes here too??

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/distrib_data_parallel.py]: The way you have it now looks good to me.

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1309]: The way you have it now looks good to me.

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/distrib_data_parallel.py]: Ok.

**Pull request merged**